detection_rules/rule_validators.py

```diff
@@ -108,6 +112,28 @@ def validate(self, data: 'QueryRuleData', meta: RuleMeta) -> None:
                 print(err_trailer)
                 raise
 
+            if not endgame_schema:
+                continue
+
+            try:
+                with endgame_schema, eql.parser.elasticsearch_syntax, eql.parser.ignore_missing_functions:
+                    eql.parse_query(self.query)
+            except eql.EqlParseError as exc:
+                message = exc.error_msg
+                trailer = err_trailer
+
+                if "Field not recognized" in message:
+                    text_fields = self.text_fields(eql_schema)
+                    if text_fields:
+                        fields_str = ', '.join(text_fields)
+                        trailer = f"\neql does not support text fields: {fields_str}\n\n{trailer}"
+
+                raise exc.__class__(exc.error_msg, exc.line, exc.column, exc.source,
+                                    len(exc.caret.lstrip()), trailer=trailer) from None
+            except Exception:
+                print(err_trailer)
+                raise
```

Comment:
Does this code seem redundant to L92-L109? Should we add a sub-function and pass the schemas in for validation?

<details>
<summary>Example Code</summary>

```
    def validate(self, data: 'QueryRuleData', meta: RuleMeta) -> None:
        """Validate an EQL query while checking TOMLRule."""
        if meta.query_schema_validation is False or meta.maturity == "deprecated":
            # syntax only, which is done via self.ast
            return

        for stack_version, mapping in meta.get_validation_stack_versions().items():
            beats_version = mapping['beats']
            ecs_version = mapping['ecs']
            endgame_version = mapping['endgame']
            err_trailer = f'stack: {stack_version}, beats: {beats_version},' \
                          f'ecs: {ecs_version}, endgame: {endgame_version}'

            beat_types, beat_schema, schema = self.get_beats_schema(data.index or [], beats_version, ecs_version)
            endgame_schema = self.get_endgame_schema(data.index, endgame_version)

            eql_schema = ecs.KqlSchema2Eql(schema)

            beat_types, beat_schema, schema = self.get_beats_schema(data.index or [], beats_version, ecs_version)
            endgame_schema = self.get_endgame_schema(data.index, endgame_version)

            eql_schema = ecs.KqlSchema2Eql(schema)

            def _validate_schema(self, schema) -> None:
                try:
                    # TODO: switch to custom cidrmatch that allows ipv6
                    with eql_schema, eql.parser.elasticsearch_syntax, eql.parser.ignore_missing_functions:
                        eql.parse_query(self.query)
                except eql.EqlParseError as exc:
                    message = exc.error_msg
                    trailer = err_trailer
                    if "Unknown field" in message and beat_types:
                        trailer = f"\nTry adding event.module or event.dataset to specify beats module\n\n{trailer}"
                    elif "Field not recognized" in message:
                        text_fields = self.text_fields(eql_schema)
                        if text_fields:
                            fields_str = ', '.join(text_fields)
                            trailer = f"\neql does not support text fields: {fields_str}\n\n{trailer}"

                    raise exc.__class__(exc.error_msg, exc.line, exc.column, exc.source,
                                        len(exc.caret.lstrip()), trailer=trailer) from None
                except Exception:
                    print(err_trailer)
                    raise

            _validate_schema(self, schema=eql_schema)
            if endgame_schema:
                _validate_schema(self, schema=endgame_schema)
```

</details>

Reply:
yup. refactored with https://github.com/elastic/detection-rules/pull/2257/commits/9759c97658fb35d2df7db734b6524a66d91c6794